Implement ionize and recombine methods in Particle class

namurphy opened the pull request:
These methods take a `Particle` object and run `__init__` again for the same element/isotope but for a different charge.  The `Particle.ionize` method will raise an `InvalidIonError` if a bare nucleus is attempted to be ionized.  The `recombine` method will issue a warning if the charge becomes negative; however, no exception will be raised in that case.

One behavior that I'm not certain about is both running `__init__` again and returning the new Particle instance.  We may wish to have just one behavior or the other.  If we have just the in-place re-initialization, then a disadvantage is that we would not be able to do comparisons like:
```Python
Particle("Fe 5+").recombine() == Particle("Fe 4+")
```
which require the new `Particle` instance to be returned.  I would appreciate more opinions on whether we should do both `__init__` and returning `self`, or just one of them.  Or, if there is even a better approach than I am taking here.  

I'd also like to check to make sure that there are no side effects from `__init__` again (e.g., if I have a list and add items to it such that every item in the list shows up twice).

Closes #287.

StanczakDominik commented on plasmapy/atomic/particle_class.py:
```diff
@@ -1328,3 +1329,136 @@ def is_ion(self) -> bool:
+        # Returning self here allows comparisons like
+        # Particle("Fe 5+").ionize() == Particle("Fe 6+")
+
+        return self
```
One could probably argue for an `inplace=True` default  argument, like Pandas does for a lot of its methods. It would be pretty simple  to include, simply wrap the two previous lines of code in an `if` statement.

namurphy replied:
Interesting!  Using the same notation as :panda_face: would be a really good idea.

namurphy commented:
I implemented @StanczakDominik's suggestion, so I'll plan on merging this soon unless there are any more suggestions.

I also rebased this to remove extraneous formatting commits and get a cleaner commit history.  I'm starting to appreciate how helpful rebasing can be in keeping the commit log clean while still being able to keep track of the commits that show the important parts of the development history and/or contain information that we may wish to remember for the future.
